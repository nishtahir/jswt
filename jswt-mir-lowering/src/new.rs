use jswt_ast::{transform::TransformVisitor, NewExpression, SingleExpression};
use jswt_common::Type;
use jswt_symbols::BindingsTable;

#[derive(Debug)]
pub struct MirNewLoweringContext<'a> {
    bindings: &'a BindingsTable,
}

impl<'a> MirNewLoweringContext<'a> {
    pub fn new(bindings: &'a BindingsTable) -> Self {
        Self { bindings }
    }
}

impl<'a> TransformVisitor for MirNewLoweringContext<'a> {
    fn visit_new(&mut self, node: &NewExpression) -> SingleExpression {
        // New expressions are desugared to a call to a constructor function.
        // visit the arguments to ensure that they are lowered if necessary
        let node_as_args = node.expression.as_arguments().expect(&format!(
            "New expressions should be followed by an arguments call"
        ));

        let args_exp = self.visit_argument_expression(node_as_args);
        let mut args_exp = args_exp.as_arguments().unwrap().clone();

        // Change the target of the arguments call to the constructor function
        let ident_exp = args_exp.ident.as_identifier_mut().unwrap();
        let ident_name = ident_exp.ident.value.clone();

        // find the class binding for the identifier
        let _ = self.bindings.lookup(&ident_name).expect(&format!(
            "Could not find class binding for identifier {}",
            ident_name
        ));

        ident_exp.ident.value = format!("{}#constructor", ident_name).into();
        args_exp.ty = Type::Binding(ident_name);
        SingleExpression::Arguments(args_exp)
    }
}

#[cfg(test)]
mod test {

    use jswt_assert::assert_debug_snapshot;
    use jswt_parser::Parser;
    use jswt_semantics::GlobalSemanticResolver;
    use jswt_symbols::ScopedSymbolTable;
    use jswt_tokenizer::Tokenizer;

    use super::*;
    use crate::MirLoweringContext;

    #[test]
    fn test_new_lowering_lowers_new_expression_into_constructor_invocation() {
        let mut tokenizer = Tokenizer::default();
        tokenizer.enqueue_source_str(
            "test_class_declaration_lowers_new_expression_into_constructor_invocation",
            r"
        class Array {
            constructor() {}
        }

        function main(): i32 {
            let x = new Array();
            return 0;
        }
    ",
        );

        let ast = Parser::new(&mut tokenizer).parse();

        let mut symbol_table = ScopedSymbolTable::default();
        let mut bindings_table = BindingsTable::default();
        let mut global_resolver =
            GlobalSemanticResolver::new(&mut bindings_table, &mut symbol_table);
        global_resolver.resolve(&ast);

        // No errors in global resolver
        assert!(global_resolver.errors().len() == 0);

        let mut lowering = MirLoweringContext::new(&bindings_table, &symbol_table);
        let lowered = lowering.lower(&ast);
        assert_debug_snapshot!(lowered);
    }
}
