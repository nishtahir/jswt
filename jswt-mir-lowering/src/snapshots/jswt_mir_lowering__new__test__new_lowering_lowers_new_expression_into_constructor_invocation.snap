---
source: jswt-mir-lowering/src/new.rs
expression: lowered
---
Ast {
    program: Program {
        files: [
            File {
                span: Span {
                    file: "test_class_declaration_lowers_new_expression_into_constructor_invocation",
                    module: "module/test_class_declaration_lowers_new_expression_into_constructor_invocation",
                    start: 9,
                    end: 163,
                },
                source_elements: SourceElements {
                    span: Span {
                        file: "test_class_declaration_lowers_new_expression_into_constructor_invocation",
                        module: "module/test_class_declaration_lowers_new_expression_into_constructor_invocation",
                        start: 9,
                        end: 163,
                    },
                    source_elements: [
                        FunctionDeclaration(
                            FunctionDeclarationElement {
                                span: Span {
                                    file: "test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                    module: "module/test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                    start: 35,
                                    end: 51,
                                },
                                decorators: FunctionDecorators {
                                    annotations: [],
                                    export: false,
                                },
                                ident: Identifier {
                                    span: Span {
                                        file: "test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                        module: "module/test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                        start: 35,
                                        end: 51,
                                    },
                                    value: "Array#constructor",
                                },
                                params: FormalParameterList {
                                    span: Span {
                                        file: "test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                        module: "module/test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                        start: 46,
                                        end: 48,
                                    },
                                    parameters: [],
                                },
                                returns: Some(
                                    TypeAnnotation {
                                        span: Span {
                                            file: "test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                            module: "module/test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                            start: 35,
                                            end: 51,
                                        },
                                        ty: Binding(
                                            "ptr",
                                        ),
                                    },
                                ),
                                body: BlockStatement {
                                    span: Span {
                                        file: "test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                        module: "module/test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                        start: 49,
                                        end: 51,
                                    },
                                    statements: StatementList {
                                        statements: [
                                            Variable(
                                                VariableStatement {
                                                    span: Span {
                                                        file: "synthetic",
                                                        module: "synthetic",
                                                        start: 0,
                                                        end: 0,
                                                    },
                                                    modifier: Const(
                                                        Span {
                                                            file: "synthetic",
                                                            module: "synthetic",
                                                            start: 0,
                                                            end: 0,
                                                        },
                                                    ),
                                                    target: Identifier(
                                                        Identifier {
                                                            span: Span {
                                                                file: "synthetic",
                                                                module: "synthetic",
                                                                start: 0,
                                                                end: 0,
                                                            },
                                                            value: "this",
                                                        },
                                                    ),
                                                    expression: Arguments(
                                                        ArgumentsExpression {
                                                            span: Span {
                                                                file: "synthetic",
                                                                module: "synthetic",
                                                                start: 0,
                                                                end: 0,
                                                            },
                                                            ident: Identifier(
                                                                IdentifierExpression {
                                                                    span: Span {
                                                                        file: "synthetic",
                                                                        module: "synthetic",
                                                                        start: 0,
                                                                        end: 0,
                                                                    },
                                                                    ident: Identifier {
                                                                        span: Span {
                                                                            file: "synthetic",
                                                                            module: "synthetic",
                                                                            start: 0,
                                                                            end: 0,
                                                                        },
                                                                        value: "malloc",
                                                                    },
                                                                    ty: Binding(
                                                                        "ptr",
                                                                    ),
                                                                },
                                                            ),
                                                            arguments: ArgumentsList {
                                                                span: Span {
                                                                    file: "synthetic",
                                                                    module: "synthetic",
                                                                    start: 0,
                                                                    end: 0,
                                                                },
                                                                arguments: [
                                                                    Literal(
                                                                        Integer(
                                                                            IntegerLiteral {
                                                                                span: Span {
                                                                                    file: "synthetic",
                                                                                    module: "synthetic",
                                                                                    start: 0,
                                                                                    end: 0,
                                                                                },
                                                                                value: 0,
                                                                                ty: Binding(
                                                                                    "ptr",
                                                                                ),
                                                                            },
                                                                        ),
                                                                    ),
                                                                ],
                                                            },
                                                            ty: Binding(
                                                                "ptr",
                                                            ),
                                                        },
                                                    ),
                                                    type_annotation: None,
                                                },
                                            ),
                                            Return(
                                                ReturnStatement {
                                                    span: Span {
                                                        file: "synthetic",
                                                        module: "synthetic",
                                                        start: 0,
                                                        end: 0,
                                                    },
                                                    expression: Identifier(
                                                        IdentifierExpression {
                                                            span: Span {
                                                                file: "test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                                                module: "module/test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                                                start: 35,
                                                                end: 51,
                                                            },
                                                            ident: Identifier {
                                                                span: Span {
                                                                    file: "test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                                                    module: "module/test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                                                    start: 35,
                                                                    end: 51,
                                                                },
                                                                value: "this",
                                                            },
                                                            ty: Binding(
                                                                "ptr",
                                                            ),
                                                        },
                                                    ),
                                                },
                                            ),
                                        ],
                                    },
                                },
                            },
                        ),
                        FunctionDeclaration(
                            FunctionDeclarationElement {
                                span: Span {
                                    file: "test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                    module: "module/test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                    start: 71,
                                    end: 158,
                                },
                                decorators: FunctionDecorators {
                                    annotations: [],
                                    export: false,
                                },
                                ident: Identifier {
                                    span: Span {
                                        file: "test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                        module: "module/test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                        start: 80,
                                        end: 84,
                                    },
                                    value: "main",
                                },
                                params: FormalParameterList {
                                    span: Span {
                                        file: "test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                        module: "module/test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                        start: 84,
                                        end: 86,
                                    },
                                    parameters: [],
                                },
                                returns: Some(
                                    TypeAnnotation {
                                        span: Span {
                                            file: "test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                            module: "module/test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                            start: 88,
                                            end: 91,
                                        },
                                        ty: Binding(
                                            "i32",
                                        ),
                                    },
                                ),
                                body: BlockStatement {
                                    span: Span {
                                        file: "test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                        module: "module/test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                        start: 92,
                                        end: 158,
                                    },
                                    statements: StatementList {
                                        statements: [
                                            Variable(
                                                VariableStatement {
                                                    span: Span {
                                                        file: "test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                                        module: "module/test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                                        start: 106,
                                                        end: 126,
                                                    },
                                                    modifier: Let(
                                                        Span {
                                                            file: "test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                                            module: "module/test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                                            start: 106,
                                                            end: 109,
                                                        },
                                                    ),
                                                    target: Identifier(
                                                        Identifier {
                                                            span: Span {
                                                                file: "test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                                                module: "module/test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                                                start: 110,
                                                                end: 111,
                                                            },
                                                            value: "x",
                                                        },
                                                    ),
                                                    expression: Arguments(
                                                        ArgumentsExpression {
                                                            span: Span {
                                                                file: "test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                                                module: "module/test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                                                start: 118,
                                                                end: 125,
                                                            },
                                                            ident: Identifier(
                                                                IdentifierExpression {
                                                                    span: Span {
                                                                        file: "test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                                                        module: "module/test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                                                        start: 118,
                                                                        end: 123,
                                                                    },
                                                                    ident: Identifier {
                                                                        span: Span {
                                                                            file: "test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                                                            module: "module/test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                                                            start: 118,
                                                                            end: 123,
                                                                        },
                                                                        value: "Array#constructor",
                                                                    },
                                                                    ty: Unknown,
                                                                },
                                                            ),
                                                            arguments: ArgumentsList {
                                                                span: Span {
                                                                    file: "test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                                                    module: "module/test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                                                    start: 123,
                                                                    end: 125,
                                                                },
                                                                arguments: [],
                                                            },
                                                            ty: Binding(
                                                                "Array",
                                                            ),
                                                        },
                                                    ),
                                                    type_annotation: None,
                                                },
                                            ),
                                            Return(
                                                ReturnStatement {
                                                    span: Span {
                                                        file: "test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                                        module: "module/test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                                        start: 146,
                                                        end: 147,
                                                    },
                                                    expression: Literal(
                                                        Integer(
                                                            IntegerLiteral {
                                                                span: Span {
                                                                    file: "test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                                                    module: "module/test_class_declaration_lowers_new_expression_into_constructor_invocation",
                                                                    start: 146,
                                                                    end: 147,
                                                                },
                                                                value: 0,
                                                                ty: Binding(
                                                                    "i32",
                                                                ),
                                                            },
                                                        ),
                                                    ),
                                                },
                                            ),
                                        ],
                                    },
                                },
                            },
                        ),
                    ],
                },
            },
        ],
    },
}
